File: website/src/features/sale-product/data/models/CashLocationModel.ts
/**
 * CashLocationModel
 * DTO and Mapper for CashLocation entity
 */

import { CashLocation, CashLocationType } from '../../domain/entities/CashLocation';

export interface CashLocationDTO {
  id: string;
  name: string;
  type: CashLocationType;
  storeId: string | null;
  isCompanyWide: boolean;
  isDeleted: boolean;
  currencyCode: string | null;
  bankAccount: string | null;
  bankName: string | null;
  locationInfo: string;
  transactionCount: number;
}

export class CashLocationModel {
  /**
   * Convert DTO to Domain Entity
   */
  static toDomain(dto: CashLocationDTO): CashLocation {
    return CashLocation.create({
      id: dto.id,
      name: dto.name,
      type: dto.type,
      storeId: dto.storeId,
      isCompanyWide: dto.isCompanyWide,
      currencyCode: dto.currencyCode,
      bankAccount: dto.bankAccount,
      bankName: dto.bankName,
    });
  }

  /**
   * Convert Domain Entity to DTO
   */
  static fromDomain(cashLocation: CashLocation): Partial<CashLocationDTO> {
    return {
      id: cashLocation.id,
      name: cashLocation.name,
      type: cashLocation.type,
      storeId: cashLocation.storeId,
      isCompanyWide: cashLocation.isCompanyWide,
      currencyCode: cashLocation.currencyCode,
      bankAccount: cashLocation.bankAccount,
      bankName: cashLocation.bankName,
    };
  }

  /**
   * Convert array of DTOs to Domain Entities
   */
  static toDomainList(dtos: CashLocationDTO[]): CashLocation[] {
    return dtos.map((dto) => CashLocationModel.toDomain(dto));
  }

  /**
   * Filter out deleted locations and convert to Domain
   */
  static filterAndConvertActive(dtos: CashLocationDTO[]): CashLocation[] {
    return dtos
      .filter((dto) => !dto.isDeleted)
      .map((dto) => CashLocationModel.toDomain(dto));
  }
}

File: myFinance_improved_V2/lib/features/sale_product/data/models/cash_location_model.dart
import '../../domain/entities/cash_location.dart';

/// Cash location model for data transfer
///
/// Supports get_cash_locations_v2 RPC response format:
/// {
///   "cash_location_id": "uuid",
///   "location_name": "Location Name",
///   "location_type": "cash" | "vault" | "bank",
///   "store_id": "uuid" | null,
///   "is_company_wide": true | false,
///   "currency_code": "VND" | null,
///   "bank_account": "..." | null,
///   "bank_name": "..." | null,
///   ...
/// }
class CashLocationModel {
  final Map<String, dynamic> json;

  CashLocationModel(this.json);

  /// Create from JSON
  factory CashLocationModel.fromJson(Map<String, dynamic> json) {
    return CashLocationModel(json);
  }

  /// Convert to domain entity
  CashLocation toEntity() {
    // v2 RPC response uses snake_case field names (actual DB column names)
    return CashLocation(
      id: json['cash_location_id']?.toString() ?? '',
      name: json['location_name']?.toString() ?? '',
      type: json['location_type']?.toString() ?? 'cash',
      storeId: json['store_id']?.toString() ?? '',
      isCompanyWide: (json['is_company_wide'] as bool?) ?? false,
      currencyCode: json['currency_code']?.toString() ?? 'VND',
      bankAccount: json['bank_account']?.toString(),
      bankName: json['bank_name']?.toString(),
    );
  }
}
